### PR TITLE
add BIP157 (Client Side Block Filtering) Messages

### PR DESCRIPTION
**Pull request by tamasblummer**

### CODE REVIEW

File: src/network/message_filter.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    pub filter_hashes: Vec<Sha256dHash>,
+}
+impl_consensus_encoding!(CFHeaders, filter_type, stop_hash, previous_filter, filter_hashes);
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+/// getcfcheckpt message
+pub struct GetCFCheckpt {
+    /// Filter type for which headers are requested
+    pub filter_type: u8,
+    /// The hash of the last block in the requested range
```
Comment (ariard) [inline src/network/message_filter.rs]: "The hash of the last block in the chain that headers are requested for"

File: src/network/message_filter.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    pub stop_hash: Sha256dHash,
+}
+impl_consensus_encoding!(GetCFCheckpt, filter_type, stop_hash);
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+/// cfcheckpt message
+pub struct CFCheckpt {
+    /// Filter type for which headers are requested
+    pub filter_type: u8,
+    /// The hash of the last block in the requested range
```
Comment (ariard) [inline src/network/message_filter.rs]: "The hash of the last block in the chain that headers are requested for"

File: src/network/message_filter.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+//!
+//! BIP157  Client Side Block Filtering network messages
+//!
```
Comment (ariard) [inline src/network/message_filter.rs]: nit: maybe add a comment a la message_blockdata: "This module describes network messages which are used for passing block filters and filter headers inside Bitcoin P2P layer"

File: src/network/message_filter.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    /// The height of the first block in the requested range
+    pub start_height: u32,
+    /// The hash of the last block in the requested range
+    pub stop_hash: Sha256dHash,
+}
+impl_consensus_encoding!(GetCFilters, filter_type, start_height, stop_hash);
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+/// cfilter message
+pub struct CFilter {
```
Comment (ariard) [inline src/network/message_filter.rs]: Btw, on cfilter message, reading the BIP, "The FilterType SHOULD match the field in the getcfilters request, and BlockHash must correspond to a block that is an ancestor of StopHash with height greater than or equal to StartHeight.". A nit, but I think there the BIP is ambiguous, BlockHash could be also a block hash that is StopHash itself so not only its ancestor ?

Comment (tamasblummer) [reply src/network/message_filter.rs]: all comments are copy-paste from BIP157 tables, not my language.

**Comment by tamasblummer** [comment]
Since all comments are copy-paste from BIP157, if they are good enough there they should be ok here. If no other issues, please approve, so I can proceed.

File: src/network/message_filter.rs (hunk 4 above)
Comment (ariard) [reply src/network/message_filter.rs]: This one isn't an issue in PR but one in the BIP itself, I was curious to have your advice on it ?

**Comment by ariard** [comment]
Oh I was just saying that some copy-paste from BIP157 seems wrong, sorry for any misunderstood, I'm not a native too :)

I don't have any rights on this repo, was just trying to clean the nits for others reviewers

File: src/network/message.rs [hunk 5]
```diff
@@ -117,7 +118,18 @@ pub enum NetworkMessage {
...
+    /// BIP157 cfilter
+    CFilter(message_filter::CFilter),
+    /// BIP157 getcfheaders
+    GetCFHeaders(message_filter::GetCFHeaders),
+    /// BIP157 cfheaders
+    CFHeaders(message_filter::CFHeaders),
+    /// BIP157 getcfcheckpt
+    GetCFCheckpt(message_filter::GetCFCheckpt),
+    /// BIP157 cfcheckpt
+    CFCheckpt(message_filter::CFCheckpt),
```
Comment (dongcarl) [inline src/network/message.rs]: What do you think about moving these to a separate enum, say... `network::message::ClientFilter`?

**Pull request merged**